devito/ir/iet/visitors.py

```diff
@@ -132,34 +131,59 @@ def _args_decl(self, args):
         """Convert an iterable of :class:`Argument` into cgen format."""
         ret = []
         for i in args:
-            if i.is_ScalarArgument:
+            if i.is_Object:
+                ret.append(c.Value('void', '*_%s' % i.name))
+            elif i.is_Scalar:
                 ret.append(c.Value('const %s' % c.dtype_to_ctype(i.dtype), i.name))
-            elif i.is_TensorArgument:
+            elif i.is_Tensor:
                 ret.append(c.Value(c.dtype_to_ctype(i.dtype),
                                    '*restrict %s_vec' % i.name))
+            elif i.is_Lowered:
+                ret.append(c.Value('const %s' % c.dtype_to_ctype(i.dtype), i.name))
             else:
                 ret.append(c.Value('void', '*_%s' % i.name))
         return ret
 
-    def _args_cast(self, args):
-        """Build cgen type casts for an iterable of :class:`Argument`."""
+    def _args_call(self, args):
+        """
+        Convert an iterable of symbols and expressions into a function
+        call signature in cgen format.
+        """
         ret = []
         for i in args:
-            if i.is_TensorArgument:
-                align = "__attribute__((aligned(64)))"
-                shape = ''.join(["[%s]" % ccode(j)
-                                 for j in i.provider.symbolic_shape[1:]])
-                lvalue = c.POD(i.dtype, '(*restrict %s)%s %s' % (i.name, shape, align))
-                rvalue = '(%s (*)%s) %s' % (c.dtype_to_ctype(i.dtype), shape,
-                                            '%s_vec' % i.name)
-                ret.append(c.Initializer(lvalue, rvalue))
-            elif i.is_PtrArgument:
-                ctype = ctypes_to_C(i.dtype)
-                lvalue = c.Pointer(c.Value(ctype, i.name))
-                rvalue = '(%s*) %s' % (ctype, '_%s' % i.name)
-                ret.append(c.Initializer(lvalue, rvalue))
+            try:
+                if i.is_Object:
+                    ret.append('*_%s' % i.name)
+                elif i.is_Array:
+                    ret.append("(%s*)%s" % (c.dtype_to_ctype(i.dtype), i.name))
+                elif i.is_Scalar:
+                    ret.append(i.name)
+                elif i.is_TensorFunction:
+                    ret.append('%s_vec' % i.name)
+            except:
```

Comment:
Something better than just except?

Reply:
Should this be addressed?

Reply:
I don't know how

Reply:
I'm guessing ``AttributeError``. Let's see...